### PR TITLE
Update offer templates

### PR DESCRIPTION
**Pull request by Henrod**

### CODE REVIEW

File: api/app.go [hunk 1]
```diff
@@ -97,15 +97,26 @@ func (a *App) getRouter() *mux.Router {
 		&VersionMiddleware{},
 	)).Methods("GET").Name("offer_templates")
 
-	r.Handle("/templates", Chain(
+	r.Handle("/templates/{id}/insert", Chain(
```
Comment (cscatolini) [inline api/app.go]: I'd keep this as POST /templates

File: api/app.go [hunk 2]
```diff
@@ -97,15 +97,26 @@ func (a *App) getRouter() *mux.Router {
...
 		&OfferTemplateHandler{App: a, Method: "insert"},
 		&NewRelicMiddleware{App: a},
 		&AuthMiddleware{App: a},
 		&LoggingMiddleware{App: a},
 		&VersionMiddleware{},
+		NewParamKeyMiddleware(a, govalidator.IsUUIDv4),
 		NewValidationMiddleware(func() interface{} { return &models.OfferTemplate{} }),
 	)).Methods("POST").Name("offer_templates")
 
+	r.HandleFunc("/templates/{id}/update", Chain(
```
Comment (cscatolini) [inline api/app.go]: Here we're creating a new version for the template with a given key, correct?

I'd use PUT /templates/{key} or something like this

File: api/offer_template.go [hunk 3]
```diff
@@ -85,6 +91,9 @@ func (g *OfferTemplateHandler) insertOfferTemplate(w http.ResponseWriter, r *htt
 	WriteBytes(w, http.StatusCreated, bytesRes)
 }
 
+func (g *OfferTemplateHandler) update(w http.ResponseWriter, r *http.Request) {
```
Comment (cscatolini) [inline api/offer_template.go]: this does nothing, you can erase it

File: models/offer.go [hunk 4]
```diff
@@ -198,12 +198,13 @@ func UpdateOfferLastSeenAt(db runner.Connection, offerID, playerID, gameID strin
 
 //GetAvailableOffers returns the offers that match the criteria of enabled offer templates
 func GetAvailableOffers(db runner.Connection, playerID, gameID string, t time.Time, mr *MixedMetricsReporter) (map[string][]*OfferToReturn, error) {
+	offerTemplatesByPlacement := make(map[string][]*OfferToReturn)
```
Comment (cscatolini) [inline models/offer.go]: 👍

File: models/offer_template.go [hunk 5]
```diff
@@ -16,7 +17,7 @@ import (
 //OfferTemplate contains the parameters of a template
 type OfferTemplate struct {
 	ID        string   `db:"id" json:"id" valid:"uuidv4"`
-	Key       string   `db:"key" json:"key" valid:"uuidv4,required"`
+	Key       string   `db:"key" json:"key" valid:"uuidv4"`
```
Comment (cscatolini) [inline models/offer_template.go]: why is key no longer required?

File: models/offer_template.go [hunk 6]
```diff
@@ -38,7 +39,7 @@ const enabledOfferTemplates = `
 //GetOfferTemplateByID returns OfferTemplate by ID
 func GetOfferTemplateByID(db runner.Connection, id string, mr *MixedMetricsReporter) (*OfferTemplate, error) {
 	var ot OfferTemplate
-	err := mr.WithDatastoreSegment("offer_templates", SegmentSelect, func() error {
+	err := mr.WithDatastoreSegment("offer_templates", "select by id", func() error {
```
Comment (cscatolini) [inline models/offer_template.go]: `SegmentSelect`

File: models/offer_template.go [hunk 7]
```diff
@@ -87,14 +88,15 @@ func GetEnabledOfferTemplates(db runner.Connection, gameID string, mr *MixedMetr
...
+
 	err = HandleNotFoundError("OfferTemplate", map[string]interface{}{"enabled": true}, err)
 	return ots, err
 }
 
 //ListOfferTemplates returns all the offer templates for a given game
 func ListOfferTemplates(db runner.Connection, gameID string, mr *MixedMetricsReporter) ([]*OfferTemplate, error) {
 	var ots []*OfferTemplate
-	err := mr.WithDatastoreSegment("offer_templates", SegmentSelect, func() error {
+	err := mr.WithDatastoreSegment("offer_templates", "select", func() error {
```
Comment (cscatolini) [inline models/offer_template.go]: `SegmentSelect`

File: models/offer_template.go [hunk 8]
```diff
@@ -109,34 +111,67 @@ func ListOfferTemplates(db runner.Connection, gameID string, mr *MixedMetricsRep
...
+func InsertOfferTemplate(db runner.Connection, ot *OfferTemplate, onlyInsert bool, mr *MixedMetricsReporter) (*OfferTemplate, error) {
+	olderOt, err := GetEnabledOfferTemplateByKeyAndGame(db, ot.Key, ot.GameID, mr)
+
+	if ot.Metadata == nil {
+		ot.Metadata = dat.JSON([]byte(`{}`))
+	}
 
 	if err != nil {
 		notFoundErr := HandleNotFoundError("OfferTemplate", map[string]interface{}{"Key": ot.Key}, err)
 		if err == notFoundErr {
```
Comment (cscatolini) [inline models/offer_template.go]: Add some comment explaining that if notFoundErr is equal to err (the error was not changed by HandleNotFoundError) it means that is is NOT a NotFoundError

File: models/offer_template.go [hunk 9]
```diff
@@ -109,34 +111,67 @@ func ListOfferTemplates(db runner.Connection, gameID string, mr *MixedMetricsRep
...
 			Returning("id, enabled").
 			QueryStruct(ot)
 	})
+	if err != nil {
+		return nil, HandleForeignKeyViolationError("OfferTemplate", err)
+	}
 
-	foreignKeyErr := HandleForeignKeyViolationError("OfferTemplate", err)
-	return ot, foreignKeyErr
+	return ot, transaction.Commit()
```
Comment (cscatolini) [inline models/offer_template.go]: assign `transaction.Commit()` to an error and then return this error

File: models/offer_template.go (hunk 7 above)
Comment (Henrod) [inline models/offer_template.go]: Oops, rebase...

File: models/offer_template.go (hunk 5 above)
Comment (Henrod) [inline models/offer_template.go]: Now I get key from the URL, so it is validated in params middleware and further added to the struct.

**Comment by cscatolini** [comment]
Closing since it will no longer be necessary.

**Pull request closed without merging**